rfcs/0004-bp-go-build-ldflags.md

```diff
@@ -0,0 +1,36 @@
+# BP_GO_BUILD_LDFLAGS
+
+## Proposal
+
+Add an environment variable `BP_GO_BUILD_LDFLAGS` which would allow users to
+set the value of `-ldflags` for their `go build` command.
+
+## Motivation
+
+The `-ldflags` flag can have complicated values with multiple sets of internal
+quotation marks, an example of this complication can be seen in [this
+issue](https://github.com/paketo-buildpacks/go-build/issues/129). The goal of
+this environment variable is to remove one quotation level hopefully making it
+easier to express the `-ldflags` value with little to no escaping.
+
+## Implementation
+
+If `BP_GO_BUILD_LDFLAGS` is set then `-ldflags` plus the value of the
+environment variable will be added to the `go build` command.
+
+The following is an example of a cumbersome `BP_GO_BUILD_FLAGS` caused by
+`-ldflags`:
+
+`BP_GO_BUILD_FLAGS="-ldflags=\"-extldflags '-f no-PIC -static'\" -tags=osusergo,netgo,embedfs"`
+
+If the same flag set were to be written using the proposed
+`BP_GO_BUILD_LDFLAGS` it would look like this:
+
+`BP_GO_BUILD_LDFLAGS="-extldflags '-f no-PIC -static'" BP_GO_BUILD_FLAGS="-tags=osusergo,netgo,embedfs"`
+
+## Unresolved Questions and Bikeshedding (Optional)
+
+* Is this necessary? Should we just document the escaping better? Should we do
+  both?
+
```

Comment:
I feel that providing users with 2 paths to do the same thing (where one is the superset of the other) might cause confusion. We'll have to deal with cases where a user sets `-ldflags` in `BP_GO_BUILD_FLAGS` and also sets a value for `BP_GO_BUILD_LDFLAGS`. Just my $0.02

Reply:
I also worry about this case. The setting of both fields seems like it would produce an ambiguous result. How do we make this reasonable for users to understand?

Reply:
The `JAVA_TOOL_OPTIONS` flag deals with an analogous issue. `JAVA_TOOL_OPTIONS` is respected by various JVMs and can be set by buildpack users at build time. But! The Java buildpacks also provide env vars like `BPL_JVM_HEAD_ROOM` that influences the value that the buildpack sets for `JAVA_TOOL_OPTIONS`. If both are set, [here's how our docs explain the precedence](https://paketo.io/docs/buildpacks/language-family-buildpacks/java/#configuring-jvm-at-runtime).  Perhaps it's sufficient to document precedence in a similar way.

I think the key question to ask is: How often will people end up setting both env vars? Seems like an uncommon or "advanced" use-case on the Java side. If it's more common to use LDFLAGS *and* other GO_FLAGS, many more "beginner" users fall into the confusing config path.

Reply:
I have added interaction examples. I hope that they help resolve some of these questions.